src/main/java/com/qudini/api/RequestSender.java

```diff
@@ -0,0 +1,490 @@
+                            .setSSLSocketFactory(sslsf)
+                            .setConnectionManager(connectionManager)
+                            // the method below is needed to remove header error with expiration date
+                            .setDefaultRequestConfig(RequestConfig.custom().setCookieSpec(CookieSpecs.STANDARD).build())
+                            .build();
+
+                    CloseableHttpResponse httpResponse = httpClient.execute(request);
+
+                    decomposedResponse = httpResponseDecomposer(httpResponse);
+
+                    httpResponse.close();
+                }
+            }
+
+
+        }catch (NoSuchAlgorithmException | IOException | KeyManagementException | KeyStoreException e){
+
+            log.error("An exception has occurred while making the intended requests, see stack trace for more details");
+
+            log.error(e.toString());
+
+        }
+
+        return decomposedResponse.get(responseAsStringKey);
+
+    }
+
+
+    private HttpPost httpPostBaseSpecification(String url, String contentType, String token){
+
+        HttpPost httpPost;
+
+        if (url.startsWith("http:") || url.startsWith("https:")) {
+
+            httpPost = new HttpPost(url);
+
+            httpPost.addHeader("Content-Type", contentType);
+
+            httpPost.addHeader("Authorization", "Basic " + token);
+
+            log.debug("Generated base http POST specification");
+
+
+        } else{
+
+            log.error(String.format("The provided URL [ %s ] does seem to be a valid URL", url));
+
+            throw new RuntimeException();
+
+        }
+
+        return httpPost;
+
+    }
+
+    private HttpPost httpPostBaseSpecification(String url, String contentType, String token, String paramsAsString, String charSet) {
+        HttpPost httpPost = httpPostBaseSpecification(url, contentType, token);
+        try{
+            httpPost.setEntity(new StringEntity(paramsAsString, charSet));
+        }catch (UnsupportedCharsetException e){
+            log.error(String.format("The charset [%s] is not valid or cannot be applied to the payload", charSet));
+
+            throw e;
+        }
+
+
+        return httpPost;
+    }
+
+    private HttpPut httpPutBaseSpecification(String url, String contentType, String token){
+
+        HttpPut httpPut;
+
+        if (url.startsWith("http:") || url.startsWith("https:")) {
+
+            httpPut = new HttpPut(url);
+
+            httpPut.addHeader("Content-Type", contentType);
+
+            httpPut.addHeader("Authorization", "Basic " + token);
+
+            log.debug("Generated base http PUT specification");
+
+
+        } else{
+
+            log.error(String.format("The provided URL [ %s ] does seem to be a valid URL", url));
+
+            throw new RuntimeException();
+
+        }
+
+        return httpPut;
+
+    }
+
+    private HttpPut httpPutBaseSpecification(String url, String contentType, String token, String paramsAsString, String charSet) {
```

Comment:
In a nutshell yes you are right, my problem is that the instance to be created is a HttpPut instead of a HttpPost and also returns the same instance.
For that I probably need to use a generic, but I could find out how. 
This comment applies to the if-else refactor

Reply:
You could returning something like https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/methods/HttpRequestBase.html or https://hc.apache.org/httpcomponents-client-ga/httpclient/apidocs/org/apache/http/client/methods/HttpEntityEnclosingRequestBase.html which are super classes of HttpPost and HttpPut.

Reply:
Ok will try to do this simplification then. Thank you to pointing a way.